src/context.rs

```diff
@@ -0,0 +1,230 @@
+use futures::channel::oneshot;
+use futures::future::{pending, select, BoxFuture, FutureExt, Shared};
+use tokio::sync::mpsc;
+use tokio::time::{delay_for, Duration};
+
+pub struct CancelHandle {
+    cancel_fn: Box<dyn FnOnce()>,
+}
+
+impl CancelHandle {
+    pub fn cancel(self) {
+        (self.cancel_fn)();
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Context {
+    pub done: Shared<BoxFuture<'static, ()>>,
+}
+
+impl Context {
+    pub fn new() -> Self {
+        Self {
+            done: pending().boxed().shared(),
+        }
+    }
+
+    pub fn with_cancel(&self) -> (Self, CancelHandle) {
+        let (cancel_tx, cancel_rx) = oneshot::channel::<()>();
+        let parent_done = self.done.clone();
+        let done = select(parent_done, cancel_rx).map(|_| ()).boxed().shared();
+        let cancel_fn = Box::new(move || {
+            let _ = cancel_tx.send(());
+        });
+        (Self { done }, CancelHandle { cancel_fn })
+    }
+
+    pub fn with_timeout(&self, timeout: Duration) -> (Self, CancelHandle) {
+        let (mut cancel_tx, mut cancel_rx) = mpsc::channel::<()>(2);
```

Comment:
I just noticed this is an `mpsc` instead of a oneshot as above in `with_cancel` even though  I think it serving the same purpose. Is there a reason for that?

Also, just a thought, but it might be nice to have the cancellable API actually be separate from the `Context` implementation, so for instance you can just write `cancellable(parent_done)` and get back the appropriate wrapped done future and corresponding `CancelHandle`. Then the implementation wouldn't be duplicated in both `with_cancel` and `with_timeout`.

Reply:
That is more or less the `abortable` API right? 

My goal for this type (`Context`) is to reduce the power of the `Future` API, and make a very explicit value that is used to control future lifecycle logic. 

`Context.done` is intended to be used always with a `select!` call. I think once we get started with workers tasks/futures you'll see where I want to go with this. I'm interested in revisiting this later as we may learn that is not really Rust idiomatic.